test/infrastructure/docker/exp/docker/nodepool.go

```diff
@@ -214,19 +217,29 @@ func (np *NodePool) refresh() error {
 		return errors.Wrapf(err, "failed to list all machines in the cluster")
 	}
 
-	np.machines = machines
+	np.machines = make([]*docker.Machine, 0, len(machines))
+	for i := range machines {
+		machine := machines[i]
+		// makes sure no control plane machines gets selected by chance.
+		if !machine.IsControlPlane() {
+			np.machines = append(np.machines, machine)
+		}
+	}
 	return nil
 }
 
 // reconcileMachine will build and provision a docker machine and update the docker machine pool status for that instance
-func (np *NodePool) reconcileMachine(ctx context.Context, machine *docker.Machine) error {
+func (np *NodePool) reconcileMachine(ctx context.Context, machine *docker.Machine, log logr.Logger) error {
 	machineStatus := getInstanceStatusByMachineName(np.dockerMachinePool, machine.Name())
 	if machineStatus == nil {
+		log.Info("Creating instance record", "instance", machine.Name())
 		machineStatus = &infrav1exp.DockerMachinePoolInstanceStatus{
 			InstanceName: machine.Name(),
 			Version:      np.machinePool.Spec.Template.Spec.Version,
 		}
 		np.dockerMachinePool.Status.Instances = append(np.dockerMachinePool.Status.Instances, machineStatus)
+		// return to surface the new machine exists.
+		return nil
```

Comment:
Seems like this should not return nil here since the machine does not currently exist and needs to be created. If returned here, a new machine would not be created / provisioned.

Reply:
I'm returning here to surface the machine exists, the machine will be bootstrapped at the next reconciliation.

If I don't return here, all the machines get bootstrapped in sequence/within the same reconciliation, and this results in the controller "freezing" for some time; with return instead, the users get a better perception that things are moving on

Reply:
What is the guarantee the docker machine pool will be requeued until all of the machines are ready? Does there need to be a check to requeue if a terminal state is not reached?

Reply:
From my test, it just gets re-queued when latest changes are picked up by the machine pool, so until you are making some progress at every reconciliation, everything is fine.

I also considered to change things and return explicitly a requeue request, but the discarded this option because the refactor was even more extensive than this PR

Reply:
This suggestion is a little overloaded, but you could return a `docker.TransientError` rather than nil. Is it really an error case, no, but it would cause a guaranteed requeue.

https://github.com/kubernetes-sigs/cluster-api/blob/46ad27b69a6abaaf30af309cc51c414c5c8da818/test/infrastructure/docker/exp/controllers/dockermachinepool_controller.go#L184-L189